services/server/src/server/services/utils/proxy-contract-util.ts

```diff
@@ -0,0 +1,112 @@
+import { whatsabi } from "@shazow/whatsabi";
+import { SourcifyChain } from "@ethereum-sourcify/lib-sourcify";
+import { AbiCoder } from "ethers";
+
+export type ProxyType =
+  | "EIP1167Proxy"
+  | "FixedProxy"
+  | "EIP1967Proxy"
+  | "GnosisSafeProxy"
+  | "DiamondProxy"
+  | "PROXIABLEProxy"
+  | "ZeppelinOSProxy"
+  | "SequenceWalletProxy";
+
+export interface ProxyDetectionResult {
+  isProxy: boolean;
+  proxyType: ProxyType | null;
+  implementations: string[];
+}
+
+export async function detectAndResolveProxy(
+  bytecode: string,
+  address: string,
+  sourcifyChain: SourcifyChain,
+): Promise<ProxyDetectionResult> {
+  // Detection
+  const codeCache = {
+    [address]: bytecode,
+  };
+  const cachedCodeProvider = whatsabi.providers.WithCachedCode(
+    {
+      request: () => {},
+    }, // Provider not needed for detection only
+    codeCache,
+  );
+
+  const detectionResult = await whatsabi.autoload(address, {
+    provider: cachedCodeProvider,
+    abiLoader: false,
+    signatureLookup: false,
+    followProxies: false,
+  });
+  const proxies = detectionResult.proxies;
+
+  // Resolution
+  const fixedProxy = proxies.find(
+    (proxy) => proxy instanceof whatsabi.proxies.FixedProxyResolver,
+  );
+  if (fixedProxy && isEIP1167Proxy(bytecode, fixedProxy.resolvedAddress)) {
+    return {
+      isProxy: true,
+      proxyType: "EIP1167Proxy",
+      implementations: [fixedProxy.resolvedAddress],
+    };
+  } // Ignore FixedProxy otherwise because it could be a false-positive
+
+  if (
+    proxies.some(
+      (proxy) => proxy instanceof whatsabi.proxies.DiamondProxyResolver,
+    )
+  ) {
+    try {
+      // Call facetAddresses()
+      const encodedFacets = await sourcifyChain.call({
+        to: address,
+        data: "0x52ef6b2c",
+      });
+      const facets = AbiCoder.defaultAbiCoder().decode(
+        ["address[]"],
+        encodedFacets,
+      )[0];
+      return {
+        isProxy: true,
+        proxyType: "DiamondProxy",
+        implementations: facets,
+      };
+    } catch (error) {
+      // Falsely detected as a diamond proxy
+    }
+  }
+
+  const checkedProxyTypes: Set<ProxyType> = new Set([
+    "FixedProxy",
+    "DiamondProxy",
+  ]); // Needed because of potential duplicates
+  for (const proxy of proxies) {
```

Comment:
Above I see we handle the EIP1167 and DiamondProxy cases. Is this here the rest of the cases for other ProxyTypes?